File: src/main/java/com/joraph/ExecutionContext.java
package com.joraph;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import java.util.function.Consumer;
import java.util.stream.Collectors;

import com.joraph.debug.DebugInfo;
import com.joraph.debug.JoraphDebug;
import com.joraph.schema.EntityDescriptor;
import com.joraph.schema.ForeignKey;
import com.joraph.schema.Property;
import com.joraph.schema.Schema;
import com.joraph.schema.SchemaUtil;
import com.joraph.schema.UnknownEntityDescriptorException;

import kotlin.jvm.functions.Function1;

/**
 * An execution context which brings together a {@link com.joraph.JoraphContext},
 * a single entity class, and the root objects.
 */
public class ExecutionContext {

	private final JoraphContext context;
	private final Schema schema;
	private final Query query;
	private final ObjectGraph objectGraph;
	private final KeysToLoad keysToLoad;

	/**
	 * Creates a new instance of ExecutionContext.
	 */
	public ExecutionContext(JoraphContext context, Query query) {
		this.context		= context;
		this.schema         = context.getSchema();
		this.query 			= query;
		this.keysToLoad		= new KeysToLoad();

		this.objectGraph = query.hasExistingGraph()
				? query.getExistingGraph()
				: new ObjectGraph(context.getSchema());
		
		addToResults(query.getRootObjects());
	}

	/**
	 * <p>Executes the plan, iterates the resulting operations, and returns the results.</p>
	 * <p>Subsequent calls to {@code execute} result in a cached {@link ObjectGraph}.</p>
	 * @return the results derived from loading the associated objects supplied in the root
	 * objects
	 */
	public ObjectGraph execute() {

		keysToLoad.clear();


		final Set<EntityDescriptor<?>> descriptors = query.getEntityClasses().stream()
				.flatMap((clazz) -> schema.getEntityDescriptors(clazz).stream())
				.collect(Collectors.toSet());


		// for each entity
		boolean keepLoading = true;
		while (keepLoading) {

			// get all of the FKs
			for (EntityDescriptor<?> desc : descriptors) {
				for (Map.Entry<Function1<?,?>, ForeignKey<?,?>> fk : desc.getForeignKeys().entrySet()) {
					gatherValuesForForeignKeysTo(fk.getValue().getForeignEntity());
				}
			}

			// figure out which entities to load based on the FKs
			Set<Class<?>> entitiesToLoad = keysToLoad.getEntitiesToLoad();

			// load the new entities
			if (entitiesToLoad.size() == 1) {
				loadEntities(entitiesToLoad.iterator().next());
			} else if (entitiesToLoad.size() > 1) {
				loadEntitiesInParallel(entitiesToLoad);
			}

			// add the new descriptors to the list
			boolean newDescriptors = descriptors.addAll(entitiesToLoad.stream()
					.flatMap((clazz) -> schema.getEntityDescriptors(clazz).stream())
					.collect(Collectors.toSet()));

			keepLoading = !entitiesToLoad.isEmpty() || newDescriptors;
		}

		JoraphDebug.addObjectGraph(objectGraph);

		return objectGraph;
	}

	private void addToResults(Iterable<?> objects) {
		if (objects == null) {
			return;
		}

		final Schema schema = Objects.requireNonNull(context.getSchema(), "Schema must not be null");

		for (Object object : objects) {
			if (object==null) {
				continue;
			}
			final EntityDescriptor<?> entityDescriptor = schema.getEntityDescriptors(object.getClass()).stream()
					.filter((d) -> d.getEntityClass().equals(object.getClass()))
					.findFirst()
					.orElseThrow(() -> new UnknownEntityDescriptorException(object.getClass()));

			final Property<?, ?> pk = entityDescriptor.getPrimaryKey();
			objectGraph.addResult(entityDescriptor.getGraphKey(), pk.read(object), object);
		}
	}

	private void gatherValuesForForeignKeysTo(Class<?> entityClass) {

		context.getSchema().getEntityDescriptors(entityClass).stream()
				.flatMap((entityDescriptor) -> context.getSchema().describeForeignKeysTo(entityClass).stream()
						.filter((fk) -> SchemaUtil.shouldLoad(fk, query.getArguments()))
						.flatMap((fk) -> objectGraph.stream(fk.getEntityClass())
								.filter((o) -> o.getClass().equals(fk.getEntityClass()))
								.map(fk::read)
								.filter(Objects::nonNull)
								.map(CollectionUtil::convertToSet)
								.flatMap(Set::stream)
								.filter(Objects::nonNull)
								.filter((id) -> !objectGraph.has(entityDescriptor.getEntityClass(), id))))
				.forEach(keysToLoad.getAddKeyToLoadFunction(entityClass));

	}

	private void loadEntities(Class<?> entityClass) {
		Set<Object> ids = keysToLoad.getKeysToLoad(entityClass);
		if (ids == null || ids.isEmpty()) {
			return;
		}

		List<?> objects = context.getLoaderContext().load(entityClass, query.getArguments(), ids);
		addToResults(objects);
		keysToLoad.addKeysLoaded(entityClass, ids);
	}

	private void loadEntitiesInParallel(Collection<Class<?>> entityClasses) {
		List<Future<?>> futures = new ArrayList<>();
		for (final Class<?> entityClass : entityClasses) {
			final DebugInfo info = JoraphDebug.getDebugInfo();
			futures.add(context.getExecutorService().submit(() -> {
				JoraphDebug.setThreadDebugInfo(info);
				loadEntities(entityClass);
				JoraphDebug.clearThreadDebugInfo();
			}));
			JoraphDebug.setThreadDebugInfo(info);
		}
		for (Future<?> future : futures) {
			try {
				future.get(context.getParallelExecutorDefaultTimeoutMillis(), TimeUnit.MILLISECONDS);
			} catch (InterruptedException | ExecutionException | TimeoutException e) {
				throw new JoraphException(e);
			}
		}
	}

	/**
	 * @return the context
	 */
	public JoraphContext getContext() {
		return context;
	}

	/**
	 * @return the query
	 */
	public Query getQuery() {
		return query;
	}

	/**
	 * @return the objectGraph
	 */
	public ObjectGraph getObjectGraph() {
		return objectGraph;
	}

	/**
	 * Simple class for managing the keys that need to be loaded.
	 */
	private static class KeysToLoad {
		private final Map<Class<?>, Set<Object>> keysToLoad = new HashMap<>();
		private final Map<Class<?>, Set<Object>> keysLoaded = new HashMap<>();

		private Consumer<Object> getAddKeyToLoadFunction(Class<?> entityClass) {
			return (id) -> addKeyToLoad(entityClass, id);
		}

		private synchronized void addKeyToLoad(Class<?> entityClass, Object id) {
			if (!getKeysLoaded(entityClass).contains(id)) {
				getKeysToLoad(entityClass).add(id);
			}
		}

		private synchronized void addKeysLoaded(Class<?> entityClass, Collection<Object> ids) {
			getKeysLoaded(entityClass).addAll(ids);
			getKeysToLoad(entityClass).removeAll(ids);
		}

		private Set<Object> getKeysToLoad(Class<?> entityClass) {
			return keysToLoad.computeIfAbsent(entityClass, __ ->
					Collections.newSetFromMap(new ConcurrentHashMap<>()));
		}

		private Set<Object> getKeysLoaded(Class<?> entityClass) {
			return keysLoaded.computeIfAbsent(entityClass, __ ->
					Collections.newSetFromMap(new ConcurrentHashMap<>()));
		}

		private synchronized void clear() {
			keysToLoad.clear();
			keysLoaded.clear();
		}

		public synchronized boolean hasKeysToLoad() {
			return !getEntitiesToLoad().isEmpty();
		}

		public synchronized Set<Class<?>> getEntitiesToLoad() {
			return keysToLoad.entrySet().stream()
					.filter((e) -> !e.getValue().isEmpty())
					.map(Map.Entry::getKey)
					.collect(Collectors.toSet());
		}
	}

}

File: src/main/kotlin/com/joraph/ExecutionContext.kt
package com.joraph

import com.joraph.debug.JoraphDebug
import com.joraph.loader.EntityLoaderContext
import com.joraph.loader.LoaderFunction
import com.joraph.schema.PrimaryKeyNullPointerException
import com.joraph.schema.Property
import com.joraph.schema.Schema
import com.joraph.schema.SchemaUtil.shouldLoad
import com.joraph.schema.UnknownEntityDescriptorException
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.Future
import java.util.concurrent.TimeUnit

/**
 * An execution context is created for every [Query] that is executed. It uses the [Schema] configured
 * by the [JoraphContext] that it was created for to coordinate [LoaderFunction]s and to execute the [Query]
 * that it was created for. Technically an [ExecutionContext] could be used more than once, but it is
 * unnecessary to do so because it is not thread-safe.
 */
open class ExecutionContext @JvmOverloads constructor(
        val context: JoraphContext,
        val query: Query,
        var maxPasses: Int = 50) {

    val schema: Schema = context.schema
    val objectGraph: ObjectGraph = query.existingGraph ?: ObjectGraph(context.schema)
    val loaderContext: EntityLoaderContext = context.loaderContext

    private val keysToLoad: KeysToLoad = KeysToLoad()

    /**
     * Executes the [Query] and returns the resulting [ObjectGraph].
     */
    open fun execute(): ObjectGraph {
        addToResults(query.rootObjects)

        keysToLoad.clear()
        val descriptors = query.entityClasses
                .flatMap { schema.getEntityDescriptors(it) }
                .toMutableSet()

        // for each entity
        var keepLoading = true
        var pass = 0
        while (keepLoading) {

            pass++
            if (pass > maxPasses) {
                throw JoraphException("Maximum passes ($maxPasses) exceeded, do you have a circular dependency?")
            }

            // get all of the FKs
            for (desc in descriptors) {
                for (fk in desc.foreignKeys) {
                    gatherValuesForForeignKeysTo(fk.foreignEntity)
                }
            }

            // figure out which entities to load based on the FKs
            val entitiesToLoad = keysToLoad.entitiesToLoad

            // load the new entities
            if (entitiesToLoad.size == 1) {
                loadEntities(entitiesToLoad.first())
            } else if (entitiesToLoad.size > 1) {
                loadEntitiesInParallel(entitiesToLoad)
            }

            // add the new descriptors to the list
            val newDescriptors = descriptors.addAll(entitiesToLoad
                    .flatMap { schema.getEntityDescriptors(it) }
                    .toSet())
            keepLoading = entitiesToLoad.isNotEmpty() || newDescriptors
        }


        JoraphDebug.addObjectGraph(objectGraph)
        return objectGraph
    }

    private fun addToResults(objects: Iterable<*>?) {
        if (objects == null) {
            return
        }

        for (obj in objects) {
            if (obj == null) {
                continue
            }

            val entityDescriptor = schema.getEntityDescriptors(obj.javaClass)
                    .firstOrNull { it.entityClass == obj.javaClass }
                    ?: throw UnknownEntityDescriptorException(obj.javaClass)

            val pk: Property<*, *> = entityDescriptor.primaryKey

            val pkValue = pk.read(obj) ?: throw PrimaryKeyNullPointerException(obj.javaClass)

            objectGraph.addResult(entityDescriptor.graphKey, pkValue, obj)
        }
    }

    private fun gatherValuesForForeignKeysTo(entityClass: Class<*>) {
        context.schema.getEntityDescriptors(entityClass)
                .flatMap { ed -> context.schema.describeForeignKeysTo(entityClass)
                        .filter { fk -> shouldLoad(fk, query.arguments) }
                        .flatMap { fk ->  objectGraph.getList(fk.entityClass)
                                .filter { o -> o!!.javaClass == fk.entityClass }
                                .mapNotNull { obj -> fk.read(obj!!) }
                                .flatMap { id -> CollectionUtil.convertToSet(id) } // because it could be a one to many
                                .filterNotNull()
                                .filter { id -> !objectGraph.has(ed.entityClass, id) }}}
                .forEach(keysToLoad.getAddKeyToLoadFunction(entityClass))
    }

    private fun loadEntities(entityClass: Class<*>) {
        val ids = keysToLoad.getKeysToLoad(entityClass)
        if (ids.isEmpty()) {
            return
        }
        val objects = loaderContext.load(entityClass, query.arguments, ids)

        addToResults(objects)
        keysToLoad.addKeysLoaded(entityClass, ids)
    }

    private fun loadEntitiesInParallel(entityClasses: Collection<Class<*>>) {
        val futures = mutableListOf<Future<*>>()
        for (entityClass in entityClasses) {
            val info = JoraphDebug.getDebugInfo()
            futures.add(context.executorService.submit {
                JoraphDebug.setThreadDebugInfo(info)
                loadEntities(entityClass)
                JoraphDebug.clearThreadDebugInfo()
            })
            JoraphDebug.setThreadDebugInfo(info)
        }
        for (future in futures) {
            try {
                future.get(context.parallelExecutorDefaultTimeoutMillis, TimeUnit.MILLISECONDS)
            } catch (e: Throwable) {
                throw JoraphException(e)
            }
        }
    }

}

/**
 * Simple class for managing the keys that need to be loaded and those
 * that have already been loaded.
 */
internal class KeysToLoad {

    private val keysToLoad: MutableMap<Class<*>, MutableSet<Any>> = mutableMapOf()
    private val keysLoaded: MutableMap<Class<*>, MutableSet<Any>> = mutableMapOf()

    fun getAddKeyToLoadFunction(entityClass: Class<*>): (Any) -> Unit = { addKeyToLoad(entityClass, it) }

    @Synchronized
    fun addKeyToLoad(entityClass: Class<*>, id: Any) {
        if (id !in getKeysLoaded(entityClass)) {
            getKeysToLoad(entityClass).add(id)
        }
    }

    @Synchronized
    fun addKeysLoaded(entityClass: Class<*>, ids: Collection<Any>) {
        getKeysLoaded(entityClass).addAll(ids)
        getKeysToLoad(entityClass).removeAll(ids)
    }

    fun getKeysToLoad(entityClass: Class<*>): MutableSet<Any> {
        return keysToLoad.computeIfAbsent(entityClass) { Collections.newSetFromMap(ConcurrentHashMap()) }
    }

    fun getKeysLoaded(entityClass: Class<*>): MutableSet<Any> {
        return keysLoaded.computeIfAbsent(entityClass) { Collections.newSetFromMap(ConcurrentHashMap()) }
    }

    @Synchronized
    fun clear() {
        keysToLoad.clear()
        keysLoaded.clear()
    }

    @get:Synchronized
    val entitiesToLoad: Set<Class<*>>
        get() = keysToLoad.entries
            .filter{ !it.value.isEmpty() }
            .map{ it.key }
            .toSet()

}
